### PR TITLE
Create fog view router test utils

### PR DESCRIPTION
**Pull request by samdealy**
### Motivation
Creates a test util module for the Fog View Router integration tests. Defines a precise drop order for the `RouterTestEnvironmenet` struct, because failure to do so makes the test fail.

### Future Work
* The next PR implements the first Fog View Router test.

### CODE REVIEW

**Comment by samdealy** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2634** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2634" target="_blank">Graphite</a> 
    * **PR #2653** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2653" target="_blank">Graphite</a> 
      * **PR #2692** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2692" target="_blank">Graphite</a> 
        * **PR #2694** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2694" target="_blank">Graphite</a> 
          * **PR #2698** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2698" target="_blank">Graphite</a>  👈
            * **PR #2375** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2375" target="_blank">Graphite</a> 
      * **PR #2672** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2672" target="_blank">Graphite</a> 
    * **PR #2627** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2627" target="_blank">Graphite</a> 
    * **PR #2617** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2617" target="_blank">Graphite</a> 
    * **PR #2660** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2660" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2698?utm_source=stack-comment).

File: fog/view/server/test-utils/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+                let config = ViewConfig {
+                    chain_id: "local".to_string(),
+                    client_responder_id: ResponderId::from_str(&store_uri.addr()).unwrap(),
+                    client_listen_uri: Store(store_uri.clone()),
+                    client_auth_token_secret: None,
+                    omap_capacity,
+                    ias_spid: Default::default(),
+                    ias_api_key: Default::default(),
+                    admin_listen_uri: Default::default(),
```
Comment (nick-mobilecoin) [inline fog/view/server/test-utils/src/lib.rs:146]: :question: If I'm understanding correctly the `FogViewRouterConfig`s `admin_listen_uri` is for admin access _only_ to the router.  This `admin_listen_uri` would be for admin access to _only_ this store instance and thus why the value isn't copied.

File: fog/view/server/test-utils/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                let ra_client =
+                    AttestClient::new(&config.ias_api_key).expect("Could not create IAS client");
+
+                let mut store = ViewServer::new(
+                    config,
+                    enclave,
+                    db,
+                    ra_client,
+                    SystemTimeProvider::default(),
+                    EpochShardingStrategy::default(),
```
Comment (nick-mobilecoin) [inline fog/view/server/test-utils/src/lib.rs:168]: :question: Why doesn't this use the `config.sharding_strategy`?

**Review by nick-mobilecoin — APPROVED** [review]

File: fog/view/server/test-utils/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+};
+use mc_util_grpc::ConnectionUriGrpcioChannel;
+use mc_util_uri::ConnectionUri;
+use std::{
+    collections::HashMap,
+    str::FromStr,
+    sync::{Arc, RwLock},
+};
+
+/// Contains helper methods used by the router integration test.
```
[the remark this answers is not in the document]
Comment (samdealy) [reply fog/view/server/test-utils/src/lib.rs:31]: Done.

File: fog/view/server/test-utils/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        SqlRecoveryDbTestContext,
+        Vec<ViewServer<SgxViewEnclave, AttestClient, SqlRecoveryDb, EpochShardingStrategy>>,
+        Arc<RwLock<HashMap<FogViewStoreUri, Arc<FogViewStoreApiClient>>>>,
+    ) {
+        let db_test_context = SqlRecoveryDbTestContext::new(logger.clone());
+        let mut store_servers = Vec::new();
+        let mut store_clients = HashMap::new();
+
+        for i in 0..store_count {
+            let db = db_test_context.get_db_instance();
```
[the remark this answers is not in the document]
Comment (samdealy) [reply fog/view/server/test-utils/src/lib.rs:124]: Decided to clone it. Done.

File: fog/view/server/test-utils/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    ) -> (
+        SqlRecoveryDbTestContext,
+        Vec<ViewServer<SgxViewEnclave, AttestClient, SqlRecoveryDb, EpochShardingStrategy>>,
+        Arc<RwLock<HashMap<FogViewStoreUri, Arc<FogViewStoreApiClient>>>>,
+    ) {
+        let db_test_context = SqlRecoveryDbTestContext::new(logger.clone());
+        let mut store_servers = Vec::new();
+        let mut store_clients = HashMap::new();
+
+        for i in 0..store_count {
```
[the remark this answers is not in the document]
Comment (samdealy) [reply fog/view/server/test-utils/src/lib.rs:123]: I don't think that applies here. 0 is a valid `block_index` for this test's purposes. It's also how the existing `smoke_tests` for fog view "classic" works.

File: fog/view/server/test-utils/src/lib.rs (hunk 1 above)
Comment (samdealy) [reply fog/view/server/test-utils/src/lib.rs:146]: yea

File: fog/view/server/test-utils/src/lib.rs (hunk 2 above)
Comment (samdealy) [reply fog/view/server/test-utils/src/lib.rs:168]: Changed.

File: fog/view/server/test-utils/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        verifier.mr_signer(mr_signer_verifier).debug(DEBUG_ENCLAVE);
+        let router_client = FogViewRouterGrpcClient::new(router_uri, verifier, grpcio_env, logger);
+
+        Self {
+            db_test_context: Some(db_test_context),
+            router_server: Some(router_server),
+            router_client: Some(router_client),
+            store_servers: Some(store_servers),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (samdealy) [reply fog/view/server/test-utils/src/lib.rs:107]: Done.

**Review by jcape — APPROVED** [review]

**Pull request merged**